modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/ParsedMatrixStats.java

```diff
@@ -0,0 +1,247 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.matrix.stats;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.ParsedAggregation;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class ParsedMatrixStats extends ParsedAggregation implements MatrixStats {
+
+    private Map<String, Long> counts;
+    private Map<String, Double> means;
+    private Map<String, Double> variances;
+    private Map<String, Double> skewness;
+    private Map<String, Double> kurtosis;
+    private Map<String, Map<String, Double>> covariances;
+    private Map<String, Map<String, Double>> correlations;
+
+    @Override
+    public String getType() {
+        return MatrixStatsAggregationBuilder.NAME;
+    }
+
+    @Override
+    public long getDocCount() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long getFieldCount(String field) {
+        if (counts == null || counts.containsKey(field) == false) {
+            return 0;
+        }
+        return counts.get(field);
+    }
+
+    @Override
+    public double getMean(String field) {
+        if (means == null) {
+            return Double.NaN;
+        }
+        return checkedGet(means, field);
+    }
+
+    @Override
+    public double getVariance(String field) {
+        if (variances == null) {
+            return Double.NaN;
+        }
+        return checkedGet(variances, field);
+    }
+
+    @Override
+    public double getSkewness(String field) {
+        if (skewness == null) {
+            return Double.NaN;
+        }
+        return checkedGet(skewness, field);
+    }
+
+    @Override
+    public double getKurtosis(String field) {
+        if (kurtosis == null) {
+            return Double.NaN;
+        }
+        return checkedGet(kurtosis, field);
+    }
+
+    @Override
+    public double getCovariance(String fieldX, String fieldY) {
+        if (covariances == null) {
+            return Double.NaN;
+        }
+        return checkedGet(checkedGet(covariances, fieldX), fieldY);
```

Comment:
nit: I'm not entirely sure this does exactly what getValFromUpperTriangularMatrix() does (e.g. I think swapping the fieldX, fieldY arguments should return the same result in the original implementation, I'm not entirely sure though). It should be possible to declare that method as static in `MatrixStatsResults` and then use it from here to get the same behaviour.